gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -410,6 +438,103 @@ ChannelCredentials createMtlsChannelCredentials() throws IOException, GeneralSec
+   *     S2A
+   */
+  ChannelCredentials createPlaintextToS2AChannelCredentials(String plaintextAddress) {
+    if (Strings.isNullOrEmpty(plaintextAddress)) {
+      return null;
+    }
+    return S2AChannelCredentials.newBuilder(plaintextAddress, InsecureChannelCredentials.create())
+        .build();
+  }
+
+  /**
+   * This method creates gRPC {@link ChannelCredentials} configured to use S2A to estbalish a mTLS
+   * connection. First, the address of S2A is discovered by using the {@link S2A} utility to learn
+   * the {@code mtlsAddress} to reach S2A and the {@code plaintextAddress} to reach S2A. Prefer to
+   * use the {@code mtlsAddress} address to reach S2A if it is non-empty and the MTLS-MDS
+   * credentials can successfully be discovered and used to create {@link TlsChannelCredentials}. If
+   * there is any failure using mTLS-to-S2A, fallback to using a plaintext connection to S2A using
+   * the {@code plaintextAddress}. If {@code plaintextAddress} is not available, this function
+   * returns null; in this case S2A will not be used, and a TLS connection to the service will be
+   * established.
+   *
+   * @return {@link ChannelCredentials} configured to use S2A to create mTLS connection to
+   *     mtlsEndpoint.
+   */
+  ChannelCredentials createS2ASecuredChannelCredentials() {
+    S2A s2aUtils = S2A.newBuilder().build();
+    String plaintextAddress = s2aUtils.getPlaintextS2AAddress();
+    String mtlsAddress = s2aUtils.getMtlsS2AAddress();
+    if (Strings.isNullOrEmpty(mtlsAddress)) {
+      // Fallback to plaintext connection to S2A.
+      return createPlaintextToS2AChannelCredentials(plaintextAddress);
+    }
+    // Currently, MTLS to MDS is only available on GCE. See:
+    // https://cloud.google.com/compute/docs/metadata/overview#https-mds
+    // Try to load MTLS-MDS creds.
+    InputStream trustBundle = null;
+    InputStream privateKey = null;
+    InputStream certChain = null;
+    try {
+      trustBundle = new FileInputStream(MTLS_MDS_ROOT);
+      privateKey = new FileInputStream(MTLS_MDS_CERT_CHAIN_AND_KEY);
+      certChain = new FileInputStream(MTLS_MDS_CERT_CHAIN_AND_KEY);
+    } catch (FileNotFoundException ignore) {
+      // Fallback to plaintext-to-S2A connection.
+      LOG.log(
+          Level.INFO,
+          "Cannot establish an mTLS connection to S2A due to error loading MTLS to MDS credentials, falling back to plaintext connection to S2A: "
+              + ignore.getMessage());
+      return createPlaintextToS2AChannelCredentials(plaintextAddress);
+    }
+    ChannelCredentials mtlsToS2AChannelCredentials = null;
+    try {
+      // Try to connect to S2A using mTLS.
+      mtlsToS2AChannelCredentials =
+          createMtlsToS2AChannelCredentials(trustBundle, privateKey, certChain);
+    } catch (IOException ignore) {
+      // Fallback to plaintext-to-S2A connection.
+      LOG.log(
+          Level.INFO,
```

Comment:
Same comment as the above `FileNotFoundException` sceanrio.

Reply:
This one, I agree that it makes sense to log as warning, since we have already verified that the credentials are available (i.e. that they can be found and opened on the file system). This `IOException` occurs when creating the mtls-to-S2A channel Credentials via gRPC's `TlsChannelCredentials`. We don't expect fallback to happen here, so would be good to log as warning that it happened.

Done in https://github.com/googleapis/sdk-platform-java/pull/3326/commits/a4ee6c5496bb0f6b9452b6e188b4ffa952e57ab3